build-tools/timing/timing.targets

```diff
@@ -63,16 +63,41 @@
     </PropertyGroup>
     <!--Simulate a clean checkout with git commands-->
     <Git
+        Condition=" '%(XACaptureBuildTimingProject.Submodule)' != 'True' "
         Arguments="rm -r --cached %(XACaptureBuildTimingProject.DirectoryToClean)"
         WorkingDirectory="$(_TopDir)"
         ToolPath="$(GitToolPath)"
-        ToolExe="$(GitToolExe)" />
+        ToolExe="$(GitToolExe)"
+    />
     <Git
+        Condition=" '%(XACaptureBuildTimingProject.Submodule)' != 'True' "
         Arguments="checkout HEAD %(XACaptureBuildTimingProject.DirectoryToClean)"
         WorkingDirectory="$(_TopDir)"
         ToolPath="$(GitToolPath)"
-        ToolExe="$(GitToolExe)" />
-    <Exec Command="$(_XABuild) /fl /flp:LogFile=$(_OutputFile).log &quot;/logger:$(_TimingLogger);ID=$(_ID);Description=$(_Description);Commit=$(XAVersionHash);OutputPath=$(_OutputFile).xml&quot; /p:Configuration=$(Configuration) /p:AndroidSupportedAbis=x86 %(XACaptureBuildTimingProject.Identity) /t:$(_Target)" />
+        ToolExe="$(GitToolExe)"
+    />
+    <!--If the XACaptureBuildTimingProject is in a submodule, need different commands-->
+    <ItemGroup>
+      <!--We should not delete these files, or will need a NuGet restore-->
+      <_GitIgnore Include="packages" />
+      <_GitIgnore Include="project.assets.json" />
+      <_GitIgnore Include="*.csproj.nuget.*" />
+    </ItemGroup>
+    <Git
+        Condition=" '%(XACaptureBuildTimingProject.Submodule)' == 'True' "
+        Arguments="reset --hard HEAD"
```

Comment:
Is this _actually_ desirable or advisable? It means that running `make run-performance-tests` will *reset* any changes done to the submodule, which may "accidentally" *blow away* any additional timing-related changes you may have made.

This seems like a Bad Idea™ for default behavior.

Reply:
This is so it can time a "fresh" build, simulating a fresh checkout & build. These are just alternative commands for submodule, note the `WorkingDirectory` on the two new commands.

We are already doing something similar for non-submodules:
https://github.com/xamarin/xamarin-android/blob/bc2b6fbf3aca0fb8303fec30724ecfe03ee4154c/build-tools/timing/timing.targets#L65-L74

Do you think there is a nicer way to do this? I could change to running the `Clean` target.

Reply:
Aside: Where do we define a `@(XACaptureBuildTimingProject)` entry with `%(XACaptureBuildTimingProject.Submodule)`=True?

```sh
$ git grep -5 '<XACaptureBuildTimingProject'
build-tools/timing/timing.projitems-<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
build-tools/timing/timing.projitems-  <PropertyGroup>
build-tools/timing/timing.projitems-    <_TopDir>$(MSBuildThisFileDirectory)..\..\</_TopDir>
build-tools/timing/timing.projitems-  </PropertyGroup>
build-tools/timing/timing.projitems-  <ItemGroup>
build-tools/timing/timing.projitems:    <XACaptureBuildTimingProject Include="$(_TopDir)samples\HelloWorld\HelloWorld.csproj">
build-tools/timing/timing.projitems-      <Name>HelloWorld</Name>
build-tools/timing/timing.projitems-      <ShortName>HelloWorld</ShortName>
build-tools/timing/timing.projitems-      <DirectoryToClean>$([System.IO.Path]::GetFullPath('$(_TopDir)samples\HelloWorld'))</DirectoryToClean>
build-tools/timing/timing.projitems-      <CSharpFile>$(_TopDir)samples\HelloWorld\MainActivity.cs</CSharpFile>
build-tools/timing/timing.projitems-      <AndroidResourceFile>$(_TopDir)samples\HelloWorld\Resources\values\Strings.xml</AndroidResourceFile>
build-tools/timing/timing.projitems-    </XACaptureBuildTimingProject>
build-tools/timing/timing.projitems:    <XACaptureBuildTimingProject Include="$(_TopDir)tests\Xamarin.Forms-Performance-Integration\Droid\Xamarin.Forms.Performance.Integration.Droid.csproj">
build-tools/timing/timing.projitems-      <Name>Xamarin.Forms-Integration</Name>
build-tools/timing/timing.projitems-      <ShortName>XF</ShortName>
build-tools/timing/timing.projitems-      <DirectoryToClean>$([System.IO.Path]::GetFullPath('$(_TopDir)tests\Xamarin.Forms-Performance-Integration'))</DirectoryToClean>
build-tools/timing/timing.projitems-      <CSharpFile>$(_TopDir)tests\Xamarin.Forms-Performance-Integration\Droid\MainActivity.cs</CSharpFile>
build-tools/timing/timing.projitems-      <AndroidResourceFile>$(_TopDir)tests\Xamarin.Forms-Performance-Integration\Droid\Resources\values\styles.xml</AndroidResourceFile>
```

My checkout doesn't have such a thing (commit f42735b3e57e0faacab6a5d830d3d8bd2f70d9ea), and this PR doesn't add such a thing, so how will these commands ever be executed?

Reply:
> Do you think there is a nicer way to do this? I could change to running the Clean target.
 
It's a matter of expectations. If I have changes in a submodule, and running a make target/MSBuild target *nukes those changes*, I'm *not* a happy camper.

(Aside: this is the *single most annoying thing* in monodroid/tools/msbuild, because *every time* I need to change androidtools, I need to edit msbuild/Makefile to *remove* the `git submodule update --init --recursive`. It angers me. Yet I still haven't fixed this yet. So who's the foolish one?)

(This is also why xamarin-android & Java.Interop have explicit `make prepare` targets, which change submodules, but *no other targets* do so. At least until this PR...)

If we *absolutely* want "fresh" timings, with *known* contents, then I think I'd prefer:

 1. *Not* using git submodules, and instead
 2. Having explicit commands to checkout the repo and init it.

Submodules that I have access to, and can inadvertently change is one thing.

Git checkouts *I don't have access to* (because we can delete the checkout once the test is done!) are "fine." They'll still make intermediate debugging annoying, but at least it's *very explicit* about what's creating the checkout, and thus what's *responsible* for the checkout (the test code).

Along this lines, I can imagine adding new `%(XACaptureBuildTimingProject.Repo)` metadata, allowing:

```xml
<XACaptureBuildTimingProject Include="SmartHotel360/src/SmartHotel.Clients/SmartHotel.Clients.Android/SmartHotel.Clients.Android.csproj">
  <Repo>https://github.com/Microsoft/SmartHotel360-mobile-desktop-apps</Repo>
  <Commit>4540a879fb9c6e6813751b849bf34871ace064de</Commit>
  <!-- ... -->
</XACaptureBuildTimingProject>
<!-- ... -->
<Git
    Condition=" '$(XACaptureBuildTimingProject.Repo)' != '' "
    Arguments="clone &quot;$(XACaptureBuildTimingProject.Repo)&quot; --depth 1 &quot;$(XACaptureBuildTimingProject.Name)&quot;"
    ...
/>
<Git
    Condition=" '$(XACaptureBuildTimingProject.Repo)' != '' "
    Arguments="checkout &quot;$(XACaptureBuildTimingProject.Commit)&quot;"
    ...
/>
```